app/components/viewer/preload.js

```diff
@@ -1,21 +1,261 @@
 const { ipcRenderer } = require('electron')
 
-const callbacks = []
+let currentId = 0.0
 
-window.read = (start, end, cb) =>
-  send('read', [start, end], (err, str) => {
-    if (err) return cb(err)
-    cb(null, Buffer.from(str, 'hex'))
+/*
+;['log', 'warn', 'error', 'info'].forEach(function (prop) {
+  const old = console[prop]
+  console[prop] = function () {
+    let arg = arguments[0]
+    if (arguments.length > 1) {
+      arg = Array.from(arguments).join(' ')
+    }
+    old.call(console, arg)
+  }
+})
+*/
+
+function nextId () {
+  currentId += 1.0
+  return currentId
+}
+
+function validateRange (file, start, end) {
+  return file.size()
+    .then(size => {
+      if (start === undefined || start === null) {
+        start = 0
+      }
+      if (end === undefined || end === null) {
+        end = size
+      }
+      if (start > size) {
+        throw new Error(`Can not create stream for ${file.datPath}[${start}:${end}] because start is out of bounds ${size}`)
+      }
+      if (end > size) {
+        throw new Error(`Can not create stream for ${file.datPath}[${start}:${end}] because end is out of bounds ${size}`)
+      }
+      if (start > end) {
+        throw new Error(`Can not create stream for ${file.datPath}[${start}:${end}] because start is after end`)
+      }
+      return {start, end}
+    })
+}
+
+class EventEmitter {
+  constructor () {
+    this.removeAllListeners()
+  }
+  removeAllListeners () {
+    this.handlers = {}
+  }
+  removeListener (event, handler) {
+    const handlers = this.handlers[event]
+    if (handlers === null || handlers === undefined) return
+    if (handlers === handler) {
+      delete this.handlers[event]
+      return
+    }
+    if (Array.isArray(handlers)) {
+      handlers.splice(handlers.findIndex(h => h === handler), 1)
+      if (handlers.length === 1) {
+        this.handlers[event] = handlers[0]
+      }
+    }
+  }
+  emit (event, ...args) {
+    const handlers = this.handlers[event]
+    // No handler
+    if (handlers === null || handlers === undefined) return
+    if (Array.isArray(handlers)) {
+      // many handlers
+      handlers.forEach(handler => handler(...args))
+      return
+    }
+    // One handler
+    handlers(...args)
+  }
+  on (event, handler) {
+    let handlers = this.handlers[event]
+    if (handlers === undefined) {
+      throw new Error(`Event ${event} not supported! (Supported events: ${Object.keys(this.handlers)})`)
+    }
+    if (Array.isArray(handlers)) {
+      handlers.push(handler)
+      return
+    }
+    if (handlers !== null) {
+      handlers = [handlers, handler]
+    } else {
+      handlers = handler
+    }
+    this.handlers[event] = handlers
+  }
+}
+
+class Stream extends EventEmitter {
+  constructor (file, start, end) {
+    super()
+    this.id = nextId()
+    this.file = file
+    this.start = start
+    this.end = end
+    this.removeAllListeners()
```

Comment:
@martinheidegger why this exists?

Reply:
true, I'm curious about this too

Reply:
In the current EventEmitter implementation `.removeAllListeners()` (re-)initializes the object and specifies what events are allowed:

https://github.com/dat-land/dat-desktop/blob/9dc26c2d3ca82adff8c4cde8b65675f91971484c/app/components/viewer/preload.js#L136-L142

and 

https://github.com/dat-land/dat-desktop/blob/9dc26c2d3ca82adff8c4cde8b65675f91971484c/app/components/viewer/preload.js#L80-L82

Reply:
It does?!

Reply:
I meant it should happen in `EventEmitter`'s constructor, not really a stream's responsibilities

Reply:
Hmm, I could pass a set of allowed events to the EventEmitter, would you be okay with that?